middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,81 @@
+
+    Yields received bytes.
+    """
+    assert isinstance(connection, socket.socket)
+    assert isinstance(encoding_expected, bool)
+
+    received_data = b''
+    message_with_payload_length = None
+    # Used to mark how many bytes was received since last decoding, to avoid decoding twice same part of data.
+    decoding_buffer = 0
+
+    while True:
+        received_data += connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        # If encoding is expected to be used by data sender,
+        # and there is encoded sequence in last received data,
+        # decode it and set encoding buffer.
+        if (
+            encoding_expected is True and (
+                ESCAPE_SEQUENCES[FRAME_SEPARATOR] in received_data[decoding_buffer:] or
+                ESCAPE_SEQUENCES[ESCAPE_CHARACTER] in received_data[decoding_buffer:]
+            )
+        ):
+            received_data = (
+                received_data[:decoding_buffer] +
+                AbstractMiddlemanMessage.escape_decode_raw_message(received_data[decoding_buffer:])
+            )
+            decoding_buffer = len(received_data)
+
+        # If separator is met, reset received data and restart.
+        if (
+            len(received_data) >= FRAME_SEPARATOR_LENGTH and
+            received_data[-FRAME_SEPARATOR_LENGTH:] == FRAME_SEPARATOR
+        ):
+            received_data = b''
+            message_with_payload_length = None
+            decoding_buffer = 0
+            continue
+
+        # When defined amount of bytes is received, read payload length.
+        # There is a range defined, because it might happen that payload length matches part of one of escape sequences,
+        # so loop should read it completely first, then decode, and then finally read payload length.
+        if FRAME_PAYLOAD_STARTING_BYTE <= len(received_data) <= (FRAME_PAYLOAD_STARTING_BYTE + len(ESCAPE_CHARACTER)):
+            # Last received data must not match start of ESCAPE_CHARACTER,
+            # otherwise ESCAPE_CHARACTER should be read completely before reading payload length.
+            if any(
+                 [ESCAPE_CHARACTER.startswith(received_data[-i:])
+                  for i in range(len(ESCAPE_CHARACTER))]
+            ):
+                continue
+
+            payload_length = AbstractMiddlemanMessage.get_frame_format().signed_part_of_the_frame.payload_length.parse(
+                received_data[FRAME_PAYLOAD_LENGTH_STARTING_BYTE:FRAME_PAYLOAD_STARTING_BYTE]
+            )
+            message_with_payload_length = FRAME_PAYLOAD_STARTING_BYTE + payload_length
+
+        # When payload length was read,
+        # and at least amount of bytes equal to payload length is received,
+        # yield received data.
+        if (
+            message_with_payload_length is not None and
+            message_with_payload_length <= len(received_data)
+        ):
+            yield received_data[:message_with_payload_length]
```

Comment:
shouldn't we "delete" the data that has been already yielded?

Reply:
Deleting is done only when `FRAME_SEPARATOR` is received. Otherwise, if for example there are extra bytes between end of payload and `FRAME_SEPARATOR`, the loop would gather them and if there is enough bytes to try to read `payload_length (`if FRAME_PAYLOAD_STARTING_BYTE <= len(received_data) <= (FRAME_PAYLOAD_STARTING_BYTE + len(ESCAPE_CHARACTER)):`) it would then set `message_with_payload_length` to some unexpected value and might try to yield message again before receiving `FRAME_SEPARATOR`.

Reply:
This function may need some comments to explain this.

Reply:
To explain stuff that the function is not doing ?